lib/pxhttpc.js

```diff
@@ -27,37 +29,71 @@ function callServer(data, headers, uri, callType, config, callback) {
-                        return callback(null, JSON.parse(data));
-                    }
-                } catch (e) {
-                    return callback('could not parse perimeterx api server response');
-                }
+
+            if (!response) {
+                return handleError(callback, S2SErrorReason.UNKNOWN_ERROR, `call to perimeterx server returned null or empty response: ${response}`, response)
             }
-            if (data) {
-                try {
-                    return callback(`perimeterx server query failed. ${JSON.parse(data).message}`);
-                } catch (e) {
-                }
+
+            if (response.statusCode === 200) {
+                return handleOkResponse(callback, response);
             }
-            return callback('perimeterx server did not return a valid response');
+            
+            return handleUnexpectedHttpResponse(callback, response);
         });
     } catch (e) {
-        return callback('error while calling perimeterx servers');
+        const s2sErrorReason = response ? S2SErrorReason.UNKNOWN_ERROR : S2SErrorReason.UNABLE_TO_SEND_REQUEST;
+        return handleError(callback, s2sErrorReason, `error while calling perimeterx servers: ${e}`, response); 
     }
 }
+
+const handleOkResponse = (callback, response) => {
+    let data = typeof response.body !== 'undefined' && response.body !== null ? response.body.toString() : null;
+
+    if (!data || data.length === 0 || data === "{}") {
+        return handleError(callback, S2SErrorReason.INVALID_RESPONSE, `unable to get data from response body ${response.body}`, response);
+    }
+    
+    if (typeof data !== 'object') {
+        try {
+            data = JSON.parse(data);
+        } catch (e) {
+            return handleError(callback, S2SErrorReason.INVALID_RESPONSE, `error parsing response body ${data}: ${e}`, response);
+        }
+    }
+    
+    if (data.status && data.status !== 0) {
+        return handleError(callback, S2SErrorReason.REQUEST_FAILED_ON_SERVER, data.message, response);
+    }
+
+    return callback(null, data);
+}
+
+const handleUnexpectedHttpResponse = (callback, response) => {
+    let errorReason = S2SErrorReason.UNKNOWN_ERROR;
+    let errorMessage = 'unexpected http status code received from perimeterx server';
+
+    if (response.statusCode >= 500) {
+        errorReason = S2SErrorReason.SERVER_ERROR;
+        errorMessage = 'perimeterx server error';
+    } else if (response.statusCode >= 400) {
+        errorReason = S2SErrorReason.BAD_REQUEST;
+        errorMessage = 'perimeterx server received bad request';
+    }
+
+    return handleError(callback, errorReason, errorMessage, response);
+}
+
+
+const handleError = (callback, errorReason, errorMessage, response) => {
+    const s2sError = response ? 
+        new S2SErrorInfo(errorReason, errorMessage, response.statusCode, response.statusMessage)
+        : new S2SErrorInfo(errorReason, errorMessage)
+    return callback(s2sError);
```

Comment:
I'm not sure what the exact flow is and what is the `callback` in this case, but does it know how to handle your new object (S2SErrorInfo)? 
because before your change it always got an explicit string.

Reply:
Yes, now the callback can handle this new type.